myco_api_client/matching_algorithms/attributed_matching_algorithm.py

```diff
@@ -21,80 +40,95 @@ class AttributedMatchingAlgorithm(BaseMatchingAlgorithm):
             cls, matching_data: Dict[str, Dict]) -> List[BidOfferMatch.serializable_dict]:
         recommendations = []
-        for market_id, data in matching_data.items():
-            bids_mapping = {bid["id"]: bid for bid in data.get("bids")}
-            offers_mapping = {offer["id"]: offer for offer in data.get("offers")}
+        for market_id, time_slot_data in matching_data.items():
+            for time_slot, data in time_slot_data.items():
+                bids_mapping = {bid["id"]: bid for bid in data.get("bids") or []}
+                offers_mapping = {offer["id"]: offer for offer in data.get("offers") or []}
 
-            if not (bids_mapping and offers_mapping):
-                continue
+                if not (bids_mapping and offers_mapping):
+                    continue
+                # Trading partners matching
+                trading_partners_recommendations = (
+                    PreferredPartnersMatchingAlgorithm.get_matches_recommendations(
+                        {market_id: data}))
+
+                bids_mapping, offers_mapping = cls._filter_out_consumed_orders(
+                    bids_mapping, offers_mapping, trading_partners_recommendations)
 
-            green_offers = cls._filter_offers_bids_by_attribute(
-                list(offers_mapping.values()), "energy_type", "PV")
-            green_bids = cls._filter_offers_bids_by_requirement(
-                list(bids_mapping.values()), "energy_type", "PV")
-            green_recommendations = PayAsBidMatchingAlgorithm.get_matches_recommendations(
-                    {market_id: {"bids": green_bids, "offers": green_offers}})
-            recommendations.extend(green_recommendations)
-            bids_mapping, offers_mapping = cls._filter_open_bids_offers(
-                bids_mapping, offers_mapping, green_recommendations)
-            residual_recommendations = PayAsBidMatchingAlgorithm.get_matches_recommendations(
-                    {market_id: {
-                        "bids": list(bids_mapping.values()),
-                        "offers": list(offers_mapping.values())}})
-            recommendations.extend(residual_recommendations)
+                # Green energy matching
+                green_recommendations = cls._perform_green_matching(
+                    market_id, offers_mapping, bids_mapping)
+
+                bids_mapping, offers_mapping = cls._filter_out_consumed_orders(
+                    bids_mapping, offers_mapping, green_recommendations)
+
+                # Residual matching
+                residual_recommendations = PayAsBidMatchingAlgorithm.get_matches_recommendations(
+                        {market_id: {
+                            "bids": bids_mapping.values(),
+                            "offers": offers_mapping.values()}})
+
+                recommendations.extend(
+                    trading_partners_recommendations
+                    + green_recommendations
+                    + residual_recommendations)
```

Comment:
As far as I see, `green_recommendations` , `residual_recommendations` and `trading_partners_recommendations` all return a dict `{market_id: [<recommendations>]}`. I assume that the intended behavior is to have a return value with `[{market_id: [<green>, <residual>, <trading_partners>]}]`. However, IIRC the + operator is not really updating the dict with the new values in case of key conflict, but uses the value of the last operand. So, for the case of conflicting keys (essentially green/residual/partner recommendations with the same market_id) only the residual recommendations would end up in the final dict. You would need to create a custom method that would merge the lists that are in the values of the green/residual etc. recommendations dicts in order for this to work as intended. 
Please correct me if I did not understand correctly the intent behind it

Reply:
@spyrostz actually all these are mere lists of dicts, not a dict
the returned values of these algorithms are lists of BIdOfferMatch dictionaries, this is where the market_id is

Reply:
Awesome! Then this will work great